internal/metrics/remote_write.go

```diff
@@ -0,0 +1,394 @@
+		requestRetryInterval = defaultRequestRetryInterval
+	}
+
+	newRemoteWrite := RemoteWrite{
+		deviceID:             deviceID,
+		tsdb:                 tsdbInstance,
+		waitInterval:         waitInterval,
+		RangeDuration:        RangeDuration,
+		RequestNumSamples:    requestNumSamples,
+		RequestRetryInterval: requestRetryInterval,
+		lastWriteFile:        path.Join(dataDir, LastWriteFileName),
+	}
+
+	lastWriteBytes, err := ioutil.ReadFile(newRemoteWrite.lastWriteFile)
+	if err == nil {
+		if len(lastWriteBytes) != 0 {
+			i, err := strconv.ParseInt(string(lastWriteBytes), 10, 64)
+			if err == nil {
+				newRemoteWrite.LastWrite = time.Unix(i/1000000000, i%1000000000)
+			} else {
+				log.Error(err)
+			}
+		}
+	} else if !os.IsNotExist(err) {
+		log.Error(err)
+	}
+
+	return &newRemoteWrite
+}
+
+// RemoteRecoverableError
+// 'remote.RecoverableError' fields are private and can't be instantiated during testing. Therefore, we use this wrapper
+type RemoteRecoverableError struct {
+	error
+}
+
+func (e RemoteRecoverableError) Error() string {
+	if e.error != nil {
+		return e.Error()
+	} else {
+		return ""
+	}
+}
+
+// WriteClient
+// an interface used for testing
+//go:generate mockgen -package=metrics -destination=write_client_mock.go . WriteClient
+type WriteClient interface {
+	Write(context.Context, []byte) error
+}
+
+type writeClient struct {
+	client remote.WriteClient
+}
+
+func (w *writeClient) Write(ctx context.Context, data []byte) error {
+	err := w.client.Store(ctx, data)
+	switch err.(type) {
+	case remote.RecoverableError:
+		return RemoteRecoverableError{err}
+	default:
+		return err
+	}
+}
+
+func (r *RemoteWrite) Update(config models.DeviceConfigurationMessage) error {
+	// the following two variables are only here until we implement reading form device configuration
+	serverURL := os.Getenv("REMOTEWRITE_SERVER_URL")
+	serverTimeout, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_SERVER_TIMEOUT"))
+	if serverTimeout == 0 {
+		serverTimeout = time.Second * 10
+	}
+
+	// create configuration
+	r.config = RemoteWriteConfiguration{
+		Url:     serverURL,
+		Timeout: serverTimeout,
+	}
+
+	if r.config.Url == "" { // feature is disabled
+		return nil
+	}
+
+	// create client
+	if r.Client == nil {
+		serverURL, err := url.Parse(r.config.Url)
+		if err != nil {
+			return err
+		}
+		client, err := remote.NewWriteClient(r.deviceID, &remote.ClientConfig{
+			Timeout: model.Duration(r.config.Timeout),
+			URL: &config_util.URL{
+				URL: serverURL,
+			},
+		})
+		if err != nil {
+			return err
+		}
+		r.Client = &writeClient{
+			client: client,
+		}
+	}
+
+	return nil
+}
+
+func (r *RemoteWrite) Init(config models.DeviceConfigurationMessage) error {
+	err := r.Update(config)
+	go r.writeRoutine()
+	return err
+}
+
+// writeRoutine
+// Used as the goroutine function for handling ongoing writes to remote server
+func (r *RemoteWrite) writeRoutine() {
+	log.Infof("metric remote writer %s started", r.deviceID)
+	for {
+		r.Write()
```

Comment:
what happens if this get stuck and no longer pushed data? Should be a timeout here with a context?

Reply:
it can not get stuck. the code is non-blocking until we reach the HTTP client. there a context with timeout is created. the timeout is configured with the server URL from device config (at the moment env var)

Reply:
well, it's a context.TODO() so no timeout at all: 

https://github.com/project-flotta/flotta-device-worker/pull/115/files/3a8d1de8e000a338a61f1e57bc29c1924ab0e1ab#diff-51602231cf99bc104842e148b00d32af4ed8812fbf41d942e7b9ed9417ee048dR334

Reply:
if u look at the Client code you'll see the timeout. Not our code.
![untitled (5)](https://user-images.githubusercontent.com/69590123/159275925-834a4a7c-9491-436e-afa8-a9f61bcd1999.png)